keps/sig-cli/0031-datadrivencommands.md

```diff
@@ -0,0 +1,412 @@
+	//    		description: "deployment namespace"
+	//
+	// +optional
+	Flags []Flag `json:"flags,omitempty"`
+
+	// SuggestFor is an array of command names for which this command will be suggested -
+	// similar to aliases but only suggests.
+	SuggestFor []string `json:"suggestFor,omitempty"`
+
+	// Aliases is an array of aliases that can be used instead of the first word in Use.
+	Aliases []string `json:"aliases,omitempty"`
+
+	// Version defines the version for this command. If this value is non-empty and the command does not
+	// define a "version" flag, a "version" boolean flag will be added to the command and, if specified,
+	// will print content of the "Version" variable.
+	// +optional
+	Version string `json:"version,omitempty"`
+}
+
+// ResourceCommandList contains a list of Commands
+type ResourceCommandList struct {
+	Items []ResourceCommand `json:"items"`
+}
+```
+
+**Example Command:**
+
+```go
+# Set Label: "cli.sigs.k8s.io/cli.v1alpha1.CommandList": ""
+# Set Annotation: "cli.sigs.k8s.io/cli.v1alpha1.CommandList": '<json>'
+---
+items:
+- command:
+    path:
+    - "create" # Command is a subcommand of this path
+    use: "deployment" # Command use
+    aliases: # Command alias'
+    - "deploy"
+    - "deployments"
+    short: Create a deployment with the specified name.
+    long: Create a deployment with the specified name.
+    example: |
+        # Create a new deployment named my-dep that runs the busybox image.
+        kubectl create deployment --name my-dep --image=busybox
+    flags:
+    - name: name
+      type: String
+      stringValue: ""
+      description: deployment name
+    - name: image
+      type: String
+      stringValue: ""
+      description: Image name to run.
+    - name: replicas
+      type: Int
+      intValue: 1
+      description: Image name to run.
+    - name: namespace
+      type: String
+      stringValue: "default"
+      description: deployment namespace
+  requests:
+  - group: apps
+    version: v1
+    resource: deployments
+    operation: Create
+    bodyTemplate: |
+      apiVersion: apps/v1
+      kind: Deployment
+      metadata:
+        name: {{index .Flags.Strings "name"}}
+        namespace: {{index .Flags.Strings "namespace"}}
+        labels:
+          app: nginx
+      spec:
+        replicas: {{index .Flags.Ints "replicas"}}
+        selector:
+          matchLabels:
+            app: {{index .Flags.Strings "name"}}
+        template:
+          metadata:
+            labels:
+              app: {{index .Flags.Strings "name"}}
+          spec:
+            containers:
+            - name: {{index .Flags.Strings "name"}}
+              image: {{index .Flags.Strings "image"}}
+    saveResponseValues:
+    - name: responsename
+      jsonPath: "{.metadata.name}"
+  output: |
+    deployment.apps/{{index .Responses.Strings "responsename"}} created
+```
+
+### Risks and Mitigations
+
+- Command name collisions: CRD publishes command that conflicts with another command
```

Comment:
Should follow the same rules we use for CRDs and api discovery priority.

Reply:
+1

Reply:
Added a comment about this